### DISCUSSION TITLE
Returning a byte array to typescript when calling a native C method in Android

### DISCUSSION THREAD

**Discussion by mrAlmond** [reactwg/react-native-new-architecture#147] [Q&A]
Hi everyone,

I'm trying to use turbo native modules to develop a RN library to encode/decode packets for a serial custom protocol.
This RN library will call an already developed C library that makes all the hard stuff about encoding/decoding.

I've already managed to make it work (at least the encoding part) in iOS.
In this case the spec file defines a method that return a **"number[]"** type.
In objective-c this is mapped by Codegen to "NSArray *" and this works correctly without big issues.

Problems come when trying to return this byte array in Android.
The Codegen maps "number[]" to WritableArray in Java, then the call to JNI requires the return value to be a jobject (I've also tried to use jbyteArray but nothing changes).

Every time I try to return a jobject  from JNI to Java the application crashes (SIGABRT).

No problems with other primitive types as jdouble or jstring.

For now the only solution is to return the string hex representation of the byte array but it's not a good solution IMHO.

Thank you for your help.

PS: I'm using RN 0.72.4.

**Comment by shamilovtim** [comment]
Hello, I would also like to piggyback on this and also support the original post. React Native seemingly lacks binary as a native JSI, TurboModule, NativeModule type that can be shared between platforms and the JS runtime. This presumably results in extra unnecessary marshaling and data massaging between binary and other types (for example, unnecessary stringifying). In this case it seems OP is successfully able to use Arrays (NSArray) to host the binary on the iOS platform, which I didn't even know was possible. However, I would argue that this should be its own type (see the table) and was necessary on the platform a long time ago. Neither NSArray or WriteableArray really seem to properly describe the use case of binary on the platform, and binary should really be its own hosted type in these tables:
<img width="480" alt="Screenshot 2023-09-05 at 1 42 58 PM" src="https://github.com/reactwg/react-native-new-architecture/assets/3495974/dd41807a-d3d8-4e63-8d55-06d762b85c53">
<img width="480" alt="Screenshot 2023-09-05 at 1 42 47 PM" src="https://github.com/reactwg/react-native-new-architecture/assets/3495974/7e5c78bb-1c4c-4278-ba53-920827e0e7db">

By the way, it seems the docs are lacking a similar table for understanding how C++ modules or TurboModules have changed these type conversions

**Reply by mrAlmond** [reply]
Hi, thank you for replying.

So this means that, for now, there is no way to return a byte array in Android from C?
The documentation also talks about ReadableArray but WritableArray is not even mentioned (but that's the type requested when using Codegen).

I've been to return a byte array in objective-c by converting che C "unsigned char *" buffer to an NSData object and then converting this to NSArray.

```
- (NSArray *)createPacket
{
    unsigned char *packetBuffer = (unsigned char *)calloc(MAX_BUF_PACK_SIZE, sizeof(unsigned char));
    if( packetBuffer == NULL )
    {
        NSLog(@"Cannot allocate packetBuffer!");
        return nil;
    }

    // This is the C method call...
    int packetSize = libcustom::libCustomCreatePacket(&packetBuffer);
    if( packetSize <= 0 )
    {
        NSLog(@"libCustom error");
        free(packetBuffer); // Free allocated buffer
        return nil;
    }
    
    // Convert the bytes array to NSData
    NSData* nsData = [NSData dataWithBytes:packetBuffer length:packetSize];
    
    free(packetBuffer); // Free allocated buffer
    
    // return NSData as a NSArray that is a type supported by react-native
    return [nsData toArray];
}
```

The "toArray" method is an extension of NSData

```
- (NSArray *)toArray { 
    const unsigned char *dataBuffer = (const unsigned char *)[self bytes];
    
    if (!dataBuffer)
        return [NSMutableArray new];
    
    NSUInteger dataLength = [self length];
    NSMutableArray *array = [NSMutableArray arrayWithCapacity:dataLength];
    
    for (int i = 0; i < dataLength; ++i)
        [array addObject:[NSNumber numberWithInteger:dataBuffer[i]]];    
    
    return array;
}
```

**Reply by shamilovtim** [reply]
@mrAlmond In my opinion the fact that we have to marshal to and from NSArray from NSData is exactly the problem! My perspective is that we should be able to go from NSData (or its Android equivalent) mapped directly to a native binary type of the JS runtime. 

As a C++ hosted object I would much prefer:
NSData -> JS Binary
JS Binary -> NSData 

As a Native Module, I'm not sure. I would hate to push the marshaling to the JS side.

**Comment by shamilovtim** [comment]
Bumping this as I run into it multiple times per year

**Comment by leontiy** [comment]
+ 1 for JS Binary <> Swift Data / NSData mapping. Is it currently possible in the new architecture? 

@shamilovtim Array can be reinterpreted as Data using `withUnsafeBytes(_:)` – it doesn't need iterating over each byte but might still involve a memcpy that wouldn't be necessary if the runtime supported the mapping) https://developer.apple.com/documentation/foundation/contiguousbytes/3139146-withunsafebytes
